### ISSUE TITLE
Support DNS challenges in the manual plugin

### ISSUE THREAD

**Issue by pde** [certbot/certbot#2236]
We have serverside and acme support for DNS challenges, at some point the manual client plugin should provide the information necessary to satisfy them.

**Comment by moshevds** [comment]
Support for DNS in the acme package is outdated, I opened PR #2461 to fix that.
The manual plugin also needs support for TLSSNI01, but there is no issue for that only a comment in the source code.

**Comment by mitchellrj** [comment]
I'm new to the project and the above commit probably doesn't conform to the high standards, but it's perhaps a useful jump-off point for this ticket. Needs tests adding and a bunch of refinement, but is ostensibly working.

**Comment by moshevds** [comment]
Hi mitchellrj,

You are among several people (myself included) that is waiting on DNS01 support. Are you willing to work on this manual plugin yourself, or did you intend this as a jump-off point for someone else? (thanks in either case)
I am not part of the letsencrypt team, but am willing to (pre-)review your code if you are looking for comments.

I have 3 comments after scanning over the code quickly:
 - Some kind of code-sharing between both (HTTP01 and DNS01) manual plugins might be a good idea. Even better if a TLSSNI01 manual mode could use that same shared code. I'm not sure if one plugin or multiple plugins would be best.
 - You are using tldextract, but this might not correspond to the actual zone configuration the user has configured. It might be better to use dns.resolver.zone_for_name, this also prevents one new dependency for the client.
 - The plugin name "dns" might be a bit ambiguous for this plugin. I am working on a RFC 2136 plugin and there are conceivably a lot more DNS plugins that could be created.

As said, if you want me to look at the code in some more detail, I am happy to do that.

**Comment by Tronic69** [comment]
Hi all,

is DNS challenge option removed from new certbot client or generally removed from letsencrypt.org?

I set it up a few weeks ago and everything worked fine until less then two weeks ago. Last time I used it on May 16th. Now it seems broken.

**Comment by bmw** [comment]
@Tronic69, the DNS challenge has never been supported by `certbot` or the `letsencrypt` client yet. The Let's Encrypt CA still supports it though.

**Comment by elee123** [comment]
@bmw, are you sure about that? I was looking around for DNS support and found this tweet in one of the GitHub threads. https://twitter.com/letsencrypt/status/689919523164721152
I can't seem to find good documentation though.

**Comment by jorhett** [comment]
The client feature is being evaluated in PR https://github.com/certbot/certbot/pull/2061

**Comment by Tronic69** [comment]
@bmw: You're right. It's not implemented direct into certbot or letsencrypt client. 
Therefore I used a script from here: https://github.com/lukas2511/letsencrypt.sh

After updating letsencrypt to certbot and updateing letsencrypt.sh everything worked again.

**Comment by brianjmurrell** [comment]
Isn't this already done with ```--manual --preferred-challenges dns-01```?

I'm more interested in it being in some automated plugin in ```certbot``` (i.e. like webroot is currently) so that I don't have to run the ```dns-01``` hook myself manually.

Is there a ticket tracking the fully automated ```dns-01``` implementation in ```certbot```?

**Comment by s-ol** [comment]
@brianjmurrell I think this should be closed too. `certbot certonly --manual --preferred-challenges dns` is exactly what OP asked for.
